### PR TITLE
Tests for new UI

### PR DESCRIPTION
**Pull request by rehandalal**
This is a starting point for tests for the new UI. It includes:
- Factories for creating properly structured but randomized test data on demand.
- Some tests for the new state tree.
- A sample test that shows how to test connected components.
- Reorganization of the state directory to match the structure of the state tree.

r?

### CODE REVIEW

File: recipe-server/client/control_new/tests/state/index.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    approvalRequests,
+    extensions,
+    recipes,
+    requests,
+    revisions,
+  },
+};
+
+
+export class UserFactory extends Factory {
```
Comment (andymikulski) [inline recipe-server/client/control_new/tests/state/index.js]: This feels out of place. Maybe we need a separate `testFactories` file.

**Comment by rehandalal** [comment]
yeah sorry, i intended to write some comments about the factories once i was finally done with implementation and totally dropped the ball. will do that now.

File: recipe-server/client/control_new/tests/state/index.js (hunk 1 above)
Comment (rehandalal) [inline recipe-server/client/control_new/tests/state/index.js]: actually what's been bugging me is that we're not normalizing users, so I'm gonna go implement that and this should be fixed.

**Comment by rehandalal** [comment]
Tagging in @mythmon just to review the factory stuff:
https://github.com/mozilla/normandy/pull/897/files#diff-e1d5dbf30b8387158fb96e3066dd6971

File: recipe-server/client/control_new/tests/factory.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+ *
+ * eg: const productData = Factory.build();
+ *
+ * New factories are created by extending this class (or it's subclasses) and defining a getFields
+ * method which returns a keyed object where the keys represent the field names and the values are
+ * an instance of the Field object or a static value to be used as is.
+ *
+ * Additionally a postGeneration function may be provided. This is always called after the data
+ * has been generated and stored in this.data. This function can be used to manipulate generated
+ * data or generate additional data based on the options provided at initialization.s
```
Comment (mythmon) [inline recipe-server/client/control_new/tests/factory.js]: Typo "at initialization.s"

File: recipe-server/client/control_new/tests/factory.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      } else if (fields[key] instanceof Field) {
+        this.data[key] = fields[key].value();
+      } else {
+        this.data[key] = fields[key];
+      }
+    });
+
+    // Check if there is a postGeneration function and if so call it.
+    if (typeof this.postGeneration === 'function') {
+      this.postGeneration();
```
Comment (mythmon) [inline recipe-server/client/control_new/tests/factory.js]: I think `postGeneration` should be defined on the base class as a no-op and let other classes override this. It simplifies this logic, and I think something about performance

**Comment by rehandalal** [comment]
@mythmon feedbacked

File: recipe-server/client/control_new/tests/factory.js (hunk 3 above)
Comment (rehandalal) [inline recipe-server/client/control_new/tests/factory.js]: I did a similar thing for `getFields`.

**Pull request merged**